module/programs/helper.js

```diff
@@ -1349,64 +1349,107 @@ module.exports = class ProgramsHelper {
-							let scope = 'scope.' + key
-							let values = userRoleInfo[key]
-							if (!Array.isArray(values)) {
-								queryFilter.push({ [scope]: { $in: values.split(',') } })
-							} else {
-								queryFilter.push({ [scope]: { $in: [...values] } })
-							}
-						})
-
-						if (data.role) {
-							queryFilter.push({
-								['scope.roles']: { $in: [messageConstants.common.ALL_ROLES, ...data.role.split(',')] },
-							})
-						}
+          // filterQuery['scope.entities'] = { $in: entities };
+          let userRoleInfo = _.omit(data, ['filter', 'factors', 'role', 'type','tenantId','orgId']);
+
+          let tenantDetails = await userService.fetchPublicTenantDetails(data.tenantId);
+					if (!tenantDetails.data || !tenantDetails.data.meta || tenantDetails.success !== true) {
+            return resolve({
+              success: false,
+              message: messageConstants.apiResponses.FAILED_TO_FETCH_TENANT_DETAILS,
+            });
+          }
+          let factors
+          if (tenantDetails.data.meta.hasOwnProperty('factors') && tenantDetails.data.meta.factors.length > 0) {
+            factors = tenantDetails.data.meta.factors;
+            let queryFilter = [];
+
+            // Build query based on each key
+            factors.forEach((factor) => {
+              let scope = 'scope.' + factor;
+              let values = userRoleInfo[factor];
+              if (!Array.isArray(values)) {
+                queryFilter.push({ [scope]: { $in: values.split(',') } });
+              } else {
+                queryFilter.push({ [scope]: { $in: [...values] } });
+              }
+            });
+            // append query filter
+            filterQuery['$and'] = queryFilter;
+          }
+          let dataToOmit = ['filter', 'role', 'factors', 'type','tenantId','orgId']
+          // factors.append(dataToOmit)
+
+          const finalKeysToRemove = [...new Set([...dataToOmit, ...factors])];
+
+          filterQuery.$or = [];
+          Object.keys(_.omit(data, finalKeysToRemove)).forEach((key) => {
+            filterQuery.$or.push({
+              [`scope.${key}`]: { $in: data[key] },
+            })
+          })
+          filterQuery['scope.entityType'] = { $in: entityTypes }
+        } else {
+          // let userRoleInfo = _.omit(data, ['filter', , 'factors', 'role','type']);
+          // let userRoleKeys = Object.keys(userRoleInfo);
+          // userRoleKeys.forEach((entities) => {
+          //   filterQuery['scope.' + entities] = {
+          //     $in: userRoleInfo[entities].split(','),
+          //   };
+          // });
+
+          // Obtain userInfo
+          let userRoleInfo = _.omit(data, ['filter', 'factors', 'role', 'type','tenantId','orgId'])
+          let userRoleKeys = Object.keys(userRoleInfo)
+          let queryFilter = []
+
+          // if factors are passed or query has to be build based on the keys passed
+          if (data.hasOwnProperty('factors') && data.factors.length > 0) {
```

Comment:
@MallanagoudaB Please make a UTILS function to modify the query based on factors. So that same logic can be shared by other functions